### ISSUE TITLE
Fails to load typeface on API 29+

### ISSUE THREAD

**Issue by ansman** [robolectric/robolectric#5881]
This is the same issue as #3590 but has a different cause. On API 29+ ResourceCompat use the `Font` APIs to load the typeface which is not shadowed it seems.

**Comment by ansman** [comment]
For now I was able to work around it like so:
```kotlin
@RunWith(RobolectricTestRunner::class)
@Config(shadows = [TypefaceCompatApi29ImplShadow::class])
class SomeTestClass {
  ...
}

@Implements(TypefaceCompatApi29Impl::class)
class TypefaceCompatApi29ImplShadow {
    companion object {
        @Implementation
        @JvmStatic
        fun createFromResourcesFontFile(
            context: Context,
            resources: Resources,
            id: Int,
            path: String,
            style: Int
        ): Typeface? =
            TypefaceCompatApi28Impl()
                .createFromResourcesFontFile(context, resources, id, path, style)
    }
}
```

**Comment by ansman** [comment]
Sure, I'm using `4.4-beta-1`

**Comment by hoisie** [comment]
Can you put up a sample app or a snippet of code that demonstrates the problem?

**Comment by ansman** [comment]
Sure, to trigger simply include a font in your project `src/main/res/font/some_font.ttf)` call `ResourcesCompat.getFont(context, R.font.some_font)` in your test while running on Android 29. It will throw the same exception as in #3590

**Comment by hoisie** [comment]
Is your test using binary resources? I.E. https://github.com/robolectric/robolectric/issues/3590#issuecomment-434778091

**Comment by hoisie** [comment]
Seems like we already have test coverage for this, and it seems to work on Android Q: https://github.com/robolectric/robolectric/pull/5888

An example app with the failure would be helpful.

**Comment by ansman** [comment]
That's interesting. I did not use binary resources but it seems to have been deprecated and removed from AGP. 
I did however try enabling this:
```kotlin
testOptions.unitTests {
    isIncludeAndroidResources = true
}
```

but it made no difference. 

My test runs just fine on API 26, 27 and 28. It's only API 29 that fails (where ResourceCompat switches implementation).

**Comment by hoisie** [comment]
What's the value of `RuntimeEnvironment.useLegacyResources()` at the start of the test?

Do you see a line something like `Called loadFromPath(/system/framework/framework-res.apk, true); mode=binary sdk=29` at the start of the test? `mode=binary` means binary resources.

Is the font a transitive dependency from a library, or is it included in the application's resource files? of the main application?

I tried to repro this on SDK 29 on a sample app and could not.

**Comment by ansman** [comment]
`RuntimeEnvironment.useLegacyResources()` is `false` in the test.

Yes, this is printed at the start of the test:
```
Called loadFromPath(/system/framework/framework-res.apk, true); mode=binary sdk=29
```

It's a transitive dependency from another gradle module that this module depends on. I tried moving the font into the module in question and it made no difference.

**Comment by hoisie** [comment]
If you run `aapt2 dump resources build/path/to/your.apk`, are you seeing the font there? Does the id match what is in your R file? It's possible you're hitting some edge case with binary resources.

**Comment by ansman** [comment]
If I build a regular APK it works just fine, and as I mentioned it works just fine on older APIs, just not 29.

When running the unit tests no APK is generated so there is nothing to inspect there.

I do see the font in `build/intermediates/res/merged/debug/font/` for the module in question after I run the tests.

**Comment by ansman** [comment]
It might be worth mentioning that both the module that includes the font and the module that has the test are library modules, not app modules.

**Comment by hoisie** [comment]
Could this be related to https://github.com/robolectric/robolectric/issues/5624? Seems like multi-module projects may have an issue using binary resources.

A fake-ish APK is generated for Robolectric tests that contains binary resources processed generated by the Android toolchain.

Is there a file named something like `apk-for-local-test.ap_` for your library module that contains the font and test?

**Comment by ansman** [comment]
Yes, I do have one of those. It contains the font and I verified that it does has the same ID as the `R.font.font_name` value has at runtime.

**Comment by hoisie** [comment]
I believe I am able to reproduce the problem now. I created an empty Android library module that contains a font, and am seeing the following exception when I try to run `ResourcesCompat.getFont(RuntimeEnvironment.application, R.font.my_font);`

```
android.content.res.Resources$NotFoundException: Font resource ID #0x7f070000 could not be retrieved.
	at androidx.core.content.res.ResourcesCompat.loadFont(ResourcesCompat.java:376)
	at androidx.core.content.res.ResourcesCompat.getFont(ResourcesCompat.java:241)
	at com.hoisie.robobugs.MainActivityTest.testFont(MainActivityTest.java:25)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:575)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:263)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by ansman** [comment]
Yup, that seems like it's the exact same issue I'm seeing!

**Comment by hoisie** [comment]
Seems to be an issue with https://github.com/robolectric/robolectric/blob/master/shadows/framework/src/main/java/org/robolectric/shadows/ShadowFontBuilder.java.

This snippet causes the same exception:
```
        new Font.Builder( RuntimeEnvironment.application.getResources(), R.font.my_font).build();
```

**Comment by hoisie** [comment]
SFTN with all the commits. This should be fixed on HEAD, but I'll try to do a 4.4.1 release that includes this fix some time this week.

**Comment by ansman** [comment]
Amazing, thanks for the quick turnaround time and for the debugging help! 🙏

**Comment by aaronweihe** [comment]
> SFTN with all the commits. This should be fixed on HEAD, but I'll try to do a 4.4.1 release that includes this fix some time this week.

is it still planned to release 4.4.1 would include this fix?

**Comment by hoisie** [comment]
4.5-alpha-1 is out with this fix.

My plan is to go directly to 4.5 and skip 4.4.1. I may be able to do 4.5
shortly as there haven't been any issues with Android R.
